common/util/type_traits.h

```diff
@@ -45,6 +45,39 @@ using match_const =
     typename std::conditional<std::is_const<Other>::value, std::add_const<T>,
                               std::remove_const<T>>::type;
 
+template <class T, class Other>
+using match_const_t = typename match_const<T, Other>::type;
+
+// Detection idiom from
+// C++ Extensions for Library Fundamentals, Version 2 (ISO/IEC TS 19568:2017)
+// https://en.cppreference.com/w/cpp/experimental/is_detected
+
+// Only utilities that have been needed so far are implemented.
+
+namespace type_traits_internal {
+template <class Default, class AlwaysVoid, template <class...> class Op,
+          class... Args>
+struct detected_impl {
+  using type = Default;
+};
+template <class Default, template <class...> class Op, class... Args>
+struct detected_impl<Default, std::void_t<Op<Args...>>, Op, Args...> {
+  using type = Op<Args...>;
+};
+
+}  // namespace type_traits_internal
+
+// Alias to `Op<Args...>` if that type is valid; otherwise alias to `Default`.
+template <class Default, template <class...> class Op, class... Args>
+using detected_or_t =
+    typename type_traits_internal::detected_impl<Default, void, Op,
+                                                 Args...>::type;
+
+// Alias to type T with removed reference, const, and volatile qualifiers.
+// Backport from C++20.
+template <class T>
+using remove_cvref_t = std::remove_cv_t<std::remove_reference_t<T>>;
```

Comment:
Is https://www.cplusplus.com/reference/type_traits/remove_cv/ enough? (C++11)

Reply:
It is not - `remove_cv` doesn't strip reference.
I've decided to backport this from C++20 because stripping of reference and const is a bit tricky in C++ and can be easily made in wrong order. With a type `T = const U&`:
- `std::remove_cv_t<std::remove_reference_t<T>>` is `U`: correct.
- `std::remove_reference_t<std::remove_cv_t<T>>` is `const U`: incorrect. `remove_cv_t` operates on a reference type itself (doing nothing), not on `const U`.